bitcoin/src/bip32.rs

```diff
@@ -561,58 +547,137 @@ impl fmt::Display for Error {
             Base58(ref e) => Some(e),
-            Hex(ref e) => Some(e),
             InvalidBase58PayloadLength(ref e) => Some(e),
-            CannotDeriveFromHardenedKey
-            | InvalidChildNumber(_)
-            | InvalidChildNumberFormat
-            | InvalidDerivationPathFormat
-            | UnknownVersion(_)
-            | WrongExtendedKeyLength(_)
-            | InvalidPublicKeyHexLength(_) => None,
+            UnknownVersion(_) | WrongExtendedKeyLength(_) => None,
             InvalidPrivateKeyPrefix => None,
             NonZeroParentFingerprintForMasterKey => None,
             NonZeroChildNumberForMasterKey => None,
         }
     }
 }
 
-impl From<secp256k1::Error> for Error {
-    fn from(e: secp256k1::Error) -> Error { Error::Secp256k1(e) }
+impl From<secp256k1::Error> for ParseError {
+    fn from(e: secp256k1::Error) -> ParseError { ParseError::Secp256k1(e) }
+}
+
+impl From<base58::Error> for ParseError {
+    fn from(err: base58::Error) -> Self { ParseError::Base58(err) }
+}
+
+impl From<InvalidBase58PayloadLengthError> for ParseError {
+    fn from(e: InvalidBase58PayloadLengthError) -> ParseError {
+        Self::InvalidBase58PayloadLength(e)
+    }
+}
+
+/// A BIP32 error
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum DerivationError {
+    /// Attempted to derive a hardened child from an xpub.
+    ///
+    /// You can only derive hardened children from xprivs.
+    CannotDeriveHardenedChild,
+    /// Attempted to derive a child of depth 256 or higher.
+    ///
+    /// There is no way to encode such xkeys.
+    MaximumDepthExceeded,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for DerivationError {}
+
+impl fmt::Display for DerivationError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Self::CannotDeriveHardenedChild =>
+                f.write_str("cannot derive hardened child of public key"),
+            Self::MaximumDepthExceeded => f.write_str("cannot derive child of depth 256 or higher"),
+        }
+    }
+}
+
+/// Out-of-range index when constructing a child number.
+///
+/// *Indices* are always in the range [0, 2^31 - 1]. Normal child numbers have the
+/// same range, while hardened child numbers lie in the range [2^31, 2^32 - 1].
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct IndexOutOfRangeError {
+    /// The index that was out of range for a child number.
+    pub index: u32,
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for IndexOutOfRangeError {}
+
+impl From<Infallible> for IndexOutOfRangeError {
+    fn from(never: Infallible) -> Self { match never {} }
+}
+
+impl fmt::Display for IndexOutOfRangeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "index {} out of range [0, 2^31 - 1] (do you have an hardened child number, rather than an index?)", self.index)
+    }
+}
+
+/// ParseError parsing a child number.
```

Comment:
```suggestion
/// Error parsing a child number.
```

This looks like a search-and-replace mistake, it is introduced in bc11589636841f105a41a0b3fe0d8e7c6a89d751